kong/pdk/log.lua

```diff
@@ -465,11 +474,172 @@ local _log_mt = {
+-- assert(kong.log.serialize().my_new_value == nil)
+--
+-- -- Dots in the key are interpreted as table accesses
+-- kong.log.set_serialize_value("my.new.value", 4)
+-- assert(kong.log.serialize().my.new_value == 4)
+--
+local set_serialize_value
+
+do
+
+  local function is_valid_value(v)
+    local t = type(v)
+    if v == nil or t == "number" or t == "string" or t == "boolean" then
+      return true
+    end
+
+    if t ~= "table" then
+      return false
+    end
+
+    for k, val in pairs(v) do
+      t = type(k)
+      if (t ~= "string"
+          and t ~= "number"
+          and t ~= "boolean")
+        or not is_valid_value(val)
+      then
+        return false
+      end
+    end
+
+    return true
+  end
+
+  set_serialize_value = function(key, value, options)
+    check_phase(phases_with_ctx)
+
+    options = options or {}
+    local mode = options.mode or "set"
+
+    if type(key) ~= "string" then
+      error("key must be a string", 2)
+    end
+
+    if not is_valid_value(value) then
+      error("value must be nil, a number, string, boolean or a table made with only numbers, string and booleans", 2)
+    end
+
+    if mode ~= "set" and mode ~= "add" and mode ~= "replace" then
+      error("mode must be 'set', 'add' or 'replace'", 2)
+    end
+
+    local ongx = (options or {}).ngx or ngx
+    local ctx = ongx.ctx
+    ctx.serialize_values = ctx.serialize_values or get_default_serialize_values()
+    ctx.serialize_values[#ctx.serialize_values + 1] =
+      { key = key, value = value, mode = mode }
+  end
+end
+
 local serialize
 
 do
-  local REDACTED_REQUEST_HEADERS = { "authorization", "proxy-authorization" }
-  local REDACTED_RESPONSE_HEADERS = {}
+  -- Modify returned table with values set with kong.log.set_serialize_values
+  local function edit_result(ctx, root)
+    local serialize_values = ctx.serialize_values or get_default_serialize_values()
+    --[[
+    ngx.log(ngx.INFO, require("inspect")({
+      a = "before",
+      serialize_values = serialize_values,
+      root = root
+    }, { newline = " ", indent = "" }))
+    ]]
+    local key, mode, new_value, subkeys, node, subkey, last_subkey, existing_value
+    for _, item in ipairs(serialize_values) do
+      key, mode, new_value = item.key, item.mode, item.value
+
+      -- Split key by ., creating subtables when needed
+      subkeys = setmetatable(split(key, "."), nil)
+      --ngx.log(ngx.INFO, require("inspect")({ subkeys = subkeys, mode = mode, new_value = new_value, key = key }, { newline = " ", indent = "" }))
+      node = root -- start in root, iterate with each subkey
+      for i = 1, #subkeys - 1 do -- note that last subkey is treated differently, below
+        subkey = subkeys[i]
+        if node[subkey] == nil then
+          if mode == "set" or mode == "add" then
+            node[subkey] = {} -- add subtables as needed
+          else
+            node = nil
+            break -- mode == replace; and we have a missing link on the "chain"
```

Comment:
should this raise an error message?

Reply:
No, that is a totally valid behavior. Consider the default:

```
{ key = "request.headers.authorization", value = "REDACTED", mode = "replace" }
```

We want to replace the header _if it exists_. But if it doesn't (or if the request has no headers whatsoever) we want to just ignore this entry and continue with the next one.